### ISSUE TITLE
Pure JS `xz` compressing/decompressing

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/module-requests#1]
> [`xz` is a lossless data compression program and file format which incorporates the LZMA/LZMA2 compression algorithms.](https://en.wikipedia.org/wiki/Xz)

The current `xz` modules on npm uses native bindings, which is very fragile and not very suitable for reusable modules. It would be nice to have a pure JS version even though it would be a lot slower. Could then make a wrapper module that used the native binding if it installed successfully, otherwise the JS version.

Might be worth looking into compiling the native [xz](https://en.wikipedia.org/wiki/XZ_Utils) code to JS using [emscripten](https://github.com/kripken/emscripten).

This could potentially also be useful in the browser, where everything *have* to be pure JS.

### Prior art

- Native binding:
  - https://www.npmjs.com/package/xz
  - https://www.npmjs.com/package/lzma-native
  - https://www.npmjs.com/package/node-liblzma
- JS
  - https://github.com/cscott/lzma-purejs (no idea if this is suitable for `xz`)

-

// @puzrin @mafintosh

**Comment by puzrin** [comment]
Personally, i'm more interested in high quality [brotli](https://github.com/google/brotli) port for ttf -> woff2 convertor.

**Comment by squarejaw** [comment]
My attempts to compile the xz code with emscripten have failed so far.  At least some of the code uses multi-threading which is [unsupported](http://kripken.github.io/emscripten-site/docs/porting/guidelines/portability_guidelines.html#code-that-cannot-be-compiled).

**Comment by sindresorhus** [comment]
@squarejaw You could compile with the `--disable-threads` flag, see the `INSTALL` file.

**Comment by squarejaw** [comment]
@sindresorhus That seems to fix part of it, but I'm still running into issues running the compiled js, e.g.
```
pipe() returning an error as we do not support them
```
If I remove some of the offending code calling `pipe()` it seems unable to find the files I specify to compress/decompress.  I'll probably dig into it some more, but after looking at the [spec](http://tukaani.org/xz/xz-file-format-1.0.4.txt) and [Java implementation](http://tukaani.org/xz/java.html) it might not be too bad to write a JS version from scratch.

**Comment by sindresorhus** [comment]
@puzrin → https://github.com/devongovett/brotli.js

**Comment by sindresorhus** [comment]
> it might not be too bad to write a JS version from scratch.

Sure, but it would most likely be slower and would require manual maintenance. Compiling it to JS with Emscripten would IMHO be the best option.

**Comment by puzrin** [comment]
@sindresorhus i know that project. It's cross-compiled, noticeable size - not interesting for me.

**Comment by puzrin** [comment]
> Sure, but it would most likely be slower and would require manual maintenance. Compiling it to JS with Emscripten would IMHO be the best option.

That depends on your goals. If you need to get result fast - cross compiling is the best choice. But size will be 4x more than manually ported. Manually ported code is not slower, if you know how to do it. We did `zlib` porting as excercise to learn v8 optimizations.

**Comment by sindresorhus** [comment]
@puzrin I need this for Node.js so size isn't really that big of an issue. Performance is though. I know you could make it as fast, but most couldn't, and it would require a lot more work than with emscripten that way.

**Comment by puzrin** [comment]
@sindresorhus you are right. In most cases, cross-compiling will be more optimal way to get predictable result in minimal time.

As i said, we had different and "crazy" (not business friendly) criterias to do zlib port - "interesting", "learning", "nobody will repeat" and so on :)

**Comment by simlu** [comment]
Any update on this?

**Comment by Richienb** [comment]
@sindresorhus Is https://www.npmjs.com/package/lzma what you're looking for?

**Comment by simlu** [comment]
@Richienb Unfortunately no. The `lzma` package also uses native bindings which exactly what the OP was trying to avoid

**Comment by pimterry** [comment]
https://www.npmjs.com/package/xzwasm was just published a couple of weeks ago, looks promising

**Comment by phanirithvij** [comment]
A side note: xzwasm only implements decompression
